### PR TITLE
Check for librsvg 2.52 and replace deprecated functions with new ones

### LINKED ISSUE #10013: Handle librsvg deprecations
Building against `librsvg 2.52.0`, the following deprecation warnings are emitted:

```
/home/sakari/builds/darktable-git/src/darktable/src/common/utility.c: In function ‘_util_get_svg_img’:
/home/sakari/builds/darktable-git/src/darktable/src/common/utility.c:422:5: warning: ‘rsvg_handle_get_dimensions’ is deprecated: Use 'rsvg_handle_get_intrinsic_size_in_pixels' instead [-Wdeprecated-declarations]
  422 |     rsvg_handle_get_dimensions(svg, &dimension);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/sakari/builds/darktable-git/src/darktable/src/common/utility.c:52:
/usr/include/librsvg-2.0/librsvg/rsvg.h:719:6: note: declared here
  719 | void rsvg_handle_get_dimensions (RsvgHandle *handle, RsvgDimensionData *dimension_data);
      |      ^~~~~~~~~~~~~~~~~~~~~~~~~~
/home/sakari/builds/darktable-git/src/darktable/src/common/utility.c:451:7: warning: ‘rsvg_handle_render_cairo’ is deprecated: Use 'rsvg_handle_render_document' instead [-Wdeprecated-declarations]
  451 |       rsvg_handle_render_cairo(svg, cr);
      |       ^~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/librsvg-2.0/librsvg/rsvg.h:1452,
                 from /home/sakari/builds/darktable-git/src/darktable/src/common/utility.c:52:
/usr/include/librsvg-2.0/librsvg/rsvg-cairo.h:95:10: note: declared here
   95 | gboolean rsvg_handle_render_cairo (RsvgHandle *handle, cairo_t *cr);
```

In fact these are turned into errors with the default build flags, so current master is unbuildable against `librsvg 2.52.0`.

### PR DESCRIPTION
**Pull request by piratenpanda**
fixes #10013

### CODE REVIEW

File: src/common/utility.c [hunk 1]
```diff
@@ -417,9 +417,18 @@ static cairo_surface_t *_util_get_svg_img(gchar *logo, const float size)
   char *dtlogo = g_build_filename(datadir, "pixmaps", logo, NULL);
   RsvgHandle *svg = rsvg_handle_new_from_file(dtlogo, &error);
   if(svg)
-  {
+  {
```
Comment (johnny-bit) [inline src/common/utility.c:420]:
```suggestion
  {
```

**Comment by flannelhead** [comment]
Thanks for doing this! As far as I understand the [migration docs](https://gitlab.gnome.org/GNOME/librsvg/-/merge_requests/588/diffs), the new APIs were actually introduced already in `librsvg 2.46` – I wonder if they should be used already starting from that version.

**Comment by piratenpanda** [comment]
I can change to 2.46 if no one objects

**Comment by flannelhead** [comment]
At least that way the new code would get more exposure and testing as many people don't probably have 2.52 yet.

**Comment by piratenpanda** [comment]
done

**Comment by piratenpanda** [comment]
So CI doesn't like 2.46 it seems

**Comment by johnny-bit** [comment]
Seems like the migration docs are incorrect then since we have this build error in CI:

```
/home/runner/work/darktable/darktable/src/src/common/utility.c: In function ‘_util_get_svg_img’:
/home/runner/work/darktable/darktable/src/src/common/utility.c:426:7: error: implicit declaration of function ‘rsvg_handle_get_intrinsic_size_in_pixels’; did you mean ‘rsvg_handle_get_intrinsic_dimensions’? [-Werror=implicit-function-declaration]
  426 |       rsvg_handle_get_intrinsic_size_in_pixels(svg, &width, &height);
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |       rsvg_handle_get_intrinsic_dimensions
compilation terminated due to -Wfatal-errors.
```

Quite honestly check the versions ;) . I have `librsvg-2.50.7` and i do not have the method.

Actually as seen in https://github.com/GNOME/librsvg/blob/69b32420e6265d50f3cfbecff36ebd9da08cd4ad/include/librsvg/rsvg.h#L970 it's been introduced in 2.52 so you have to sprinkle or correct for that ;)

**Comment by flannelhead** [comment]
> Seems like the migration docs are incorrect then since we have this build error in CI:

Probably just a misunderstanding of mine. Sorry about the hassle :/

**Comment by piratenpanda** [comment]
I think the new size determination needs 2.52 and print document 2.46. Shall we discriminate for the two methods or should I leave it at 2.52 for both?

**Comment by johnny-bit** [comment]
If you want to be correct you can try checking versions from which given methods were introduced. That way the coverage would be greater and would cover what @flannelhead said - that most don't have 2.52 yet but have newer funcions. However current ver is good as-is and it's up to you if you want to play details ;)

**Comment by piratenpanda** [comment]
rsvg_handle_render_cairo seems to be deprecated in 2.52 as well and not in 2.46 so I think the PR is fine as is: https://github.com/GNOME/librsvg/blob/69b32420e6265d50f3cfbecff36ebd9da08cd4ad/include/librsvg/rsvg-cairo.h#L91

**Comment by ralfbrown** [comment]
I would suggest refactoring the changes into two functions which each get called from two locations.

**Comment by piratenpanda** [comment]
> I would suggest refactoring the changes into two functions which each get called from two locations.

yeah, i also don't like the doubled code but I am afraid my C skills are not good enough. I thought it would be good to have functions like the following where the first returns the dimension variable which can then be used like it was before. 

```
RsvgDimensionData getSvgDimension(RsvgHandle svg) {
  RsvgDimensionData dimension;
  // rsvg_handle_get_dimensions has been deprecated in librsvg 2.52
  #if LIBRSVG_CHECK_VERSION(2,52,0)
    double width;
    double height;
    rsvg_handle_get_intrinsic_size_in_pixels(&svg, &width, &height);
    dimension.width = width;
    dimension.height = height;
  #else      
    rsvg_handle_get_dimensions(&svg, &dimension);
  #endif
  return dimension; 
}
```
while the first works I have no idea how to a) pass the cairo surface to the second function. Such thing never happens in the code so I am not sure if this is the right way to to do and b) adding svg stuff to utilities.c and .h seems like the proper place, still seems a bit strange to me. Input very much welcome.
```
void dt_render_svg(RsvgHandle svg, cairo_surface_t cr, double width, double height, double offset_x, double offset_y) {
  // rsvg_handle_render_cairo has been deprecated in librsvg 2.52
  #if LIBRSVG_CHECK_VERSION(2,52,0)
    RsvgRectangle viewport;
	  viewport.x = offset_x;
	  viewport.y = offset_y;
	  viewport.width = width;
	  viewport.height = height;
    rsvg_handle_render_document(&svg, &cr, &viewport, &error);
  #else
    rsvg_handle_render_cairo(&svg, &cr);
  #endif  
}
```

**Comment by elstoc** [comment]
I think this might need to be added to the 3.6.x branch too, as it's going to cause issues for anyone compiling 3.6 on ArchLinux.

**Comment by ralfbrown** [comment]
I would just pass in pointers to the dimension structure and cairo surface.
The second function would then beome
```
void dt_render_svg(RsvgHandle svg, cairo_surface_t &cr, double width, double height, double offset_x, double offset_y) {
  // rsvg_handle_render_cairo has been deprecated in librsvg 2.52
  #if LIBRSVG_CHECK_VERSION(2,52,0)
    RsvgRectangle viewport;
	  viewport.x = offset_x;
	  viewport.y = offset_y;
	  viewport.width = width;
	  viewport.height = height;
    rsvg_handle_render_document(&svg, cr, &viewport, &error);
  #else
    rsvg_handle_render_cairo(&svg, cr);
  #endif  
}
```

File: src/iop/watermark.c [hunk 2]
```diff
@@ -912,7 +921,19 @@ void process(struct dt_iop_module_t *self, dt_dev_pixelpipe_iop_t *piece, const
     case DT_WTM_SVG:
       cairo_scale(cr_two, scale, scale);
       /* render svg into surface*/
-      rsvg_handle_render_cairo(svg, cr_two);
+      
+      // rsvg_handle_render_cairo has been deprecated in librsvg 2.52
+      #if LIBRSVG_CHECK_VERSION(2,52,0)
+        GError *error = NULL;
+        RsvgRectangle viewport;
```
Comment (TurboGit) [inline src/iop/watermark.c:928]:
```
= { .x = 0,
    .y = 0,
    ... }:
```

File: src/common/utility.c [hunk 3]
```diff
@@ -448,7 +457,19 @@ static cairo_surface_t *_util_get_svg_img(gchar *logo, const float size)
     {
       cairo_t *cr = cairo_create(surface);
       cairo_scale(cr, factor, factor);
-      rsvg_handle_render_cairo(svg, cr);
+      
+      // rsvg_handle_render_cairo has been deprecated in librsvg 2.52
+      #if LIBRSVG_CHECK_VERSION(2,52,0)
+        RsvgRectangle viewport;
```
Comment (TurboGit) [inline src/common/utility.c:463]:
```
= { .x = 0,
    .y = 0,
    ... }:

```

**Review by TurboGit — CHANGES_REQUESTED** [review]
A little code clean proposed.

**Comment by piratenpanda** [comment]
So I changed the second function to:
```
void dt_render_svg(RsvgHandle svg, cairo_surface_t &cr, double width, double height, double offset_x, double offset_y) 
{
  // rsvg_handle_render_cairo has been deprecated in librsvg 2.52
  #if LIBRSVG_CHECK_VERSION(2,52,0)
    RsvgRectangle viewport = { 
      .x = 0,
      .y = 0,
      .width = width,
      .height = height,
    };
    rsvg_handle_render_document(&svg, cr, &viewport, &error);
  #else
    rsvg_handle_render_cairo(&svg, cr);
  #endif  
}
```

in the code the surface is created with:
```
cairo_t *cr = cairo_create(surface);
```
and afterwards I call the function with:
```
dt_render_svg(*svg, *cr, final_width, final_height, 0, 0);
```
The compiler now complains about a "invalid use of the incomplete typedef »cairo_t« {alias »struct _cairo«}"

I guess it will be a small change but I am lost where to look. I tried to look for the problem and found some inheritance things but I'm not getting any wiser here

**Comment by johnny-bit** [comment]
you want `cairo_surface_t` not `cairo_t` ;)

**Comment by ralfbrown** [comment]
`dt_render_svg(*svg, cr, final_width, final_height, 0, 0);` because you already have `cr` as a pointer and are passing a pointer rather than a structure.

And, my bad, the function header should be
```
void dt_render_svg(RsvgHandle svg, cairo_surface_t *cr, double width, double height, double offset_x, double offset_y) 
```
(My C++ background slipped through.)  I'm surprised my original proposal compiled as straight C code.

**Comment by piratenpanda** [comment]
So now it compiles but dt segfaults on startup. So I guess I missed something still. I'll take a look when I get back from work, maybe someone spots the obvious mistake earlier

**Comment by parafin** [comment]
You're copying RsvgHandle this way, which I'm not sure is allowed. You should pass RsvgHandle* as a function parameter, not RsvgHandle.

**Comment by piratenpanda** [comment]
Thank you! That was indeed the case. It's working now. This is the result of me improvising in C..

File: src/common/utility.h [hunk 4]
```diff
@@ -97,6 +98,12 @@ void dt_copy_file(const char *src, const char *dst);
 // copy the contents of a file in dt's data directory to a new file
 void dt_copy_resource_file(const char *src, const char *dst);
 
+// returns the RsvgDimensionData of a supplied RsvgHandle
+RsvgDimensionData getSvgDimension(RsvgHandle *svg);
```
Comment (TurboGit) [inline src/common/utility.h:102]: Name should be: `dt_get_svg_dimension`.

**Review by TurboGit — CHANGES_REQUESTED** [review]
dt naming issue.

**Review by TurboGit — APPROVED** [review]
Looks ok now, thanks !

**Comment by piratenpanda** [comment]
I'll prepare another pr for 3.6 branch tomorrow if that is needed?

**Comment by TurboGit** [comment]
> I'll prepare another pr for 3.6 branch tomorrow if that is needed?

No, not needed I do not expect more changes to this branch and no more 3.6.x release.

**Pull request merged**